### PR TITLE
(feat) O3-3052 Support date-options within date input component

### PR DESCRIPTION
**Pull request by jabahum**
## Requirements

- [ ] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary

This PR allows for retrospective entry of form fields by backdating the date

## Screenshots
- Disabled date if question value has not been entered
Screenshot 2024-05-14 at 15 15 00

- Recording
https://www.loom.com/share/d9fa21f0b98e444d86b4ccfa6a11b795
## Related Issue

https://openmrs.atlassian.net/browse/O3-3052

### CODE REVIEW

File: src/components/inputs/date/date.component.tsx [hunk 1]
```diff
@@ -55,7 +64,13 @@ const DateField: React.FC<FormFieldProps> = ({ question, onChange, handler, prev
       setFieldValue(question.id, refinedDate);
       onChange(question.id, refinedDate, setErrors, setWarnings);
       onTimeChange(false, true);
-      question.value = handler?.handleFieldSubmission(question, refinedDate, encounterContext);
+      question.value =
+        obsDate === undefined
```
Comment (pirupius) [inline src/components/inputs/date/date.component.tsx:68]: similar comments apply

File: src/components/inputs/date/date.component.tsx [hunk 2]
```diff
@@ -205,6 +220,18 @@ const DateField: React.FC<FormFieldProps> = ({ question, onChange, handler, prev
             ''
           )}
         </div>
+        {question.questionOptions.showDate === 'true' ? (
+          <div style={{ marginTop: '5px' }}>
+            <InlineDate
+              question={question}
+              setObsDateTime={(value) => setObsDate(value)}
+              onChange={() => {}}
```
Comment (pirupius) [inline src/components/inputs/date/date.component.tsx:228]: ditto

**Comment by samuelmale** [comment]
@jabahum did you see the merge conflicts?

File: src/components/inputs/content-switcher/content-switcher.component.tsx [hunk 3]
```diff
@@ -78,6 +83,17 @@ const ContentSwitcher: React.FC<FormFieldProps> = ({ question, onChange, handler
...
+
+        {question.questionOptions.showDate === 'true' ? (
+          <div style={{ marginTop: '5px' }}>
+            <InlineDate
+              question={question}
+              setObsDateTime={(value) => setObsDate(value)}
+            />
+          </div>
+        ) : (
+          ''
```
Comment (arodidev) [inline src/components/inputs/content-switcher/content-switcher.component.tsx:95]:
```suggestion
          null
```

Comment (arodidev) [reply src/components/inputs/content-switcher/content-switcher.component.tsx:95]: We should prefer using `null`, I believe `''` would still create a DOM node containing an empty string

File: src/components/inputs/file/file.component.tsx [hunk 4]
```diff
@@ -178,6 +181,16 @@ const File: React.FC<FileProps> = ({ question, handler }) => {
...
       )}
+      {question.questionOptions.showDate === 'true' ? (
+          <div style={{ marginTop: '5px' }}>
+            <InlineDate
+              question={question}
+              setObsDateTime={(value) => setObsDate(value)}
+            />
+          </div>
+        ) : (
+          ''
```
Comment (arodidev) [inline src/components/inputs/file/file.component.tsx:192]: Same applies here, and the rest.

File: src/components/inputs/inline-date/inline-date.component.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      <div>
+        <DatePicker
+          datePickerType="single"
+          onChange={onDateChange}
+          className={`${styles.boldedLabel} ${isFieldRequiredError ? styles.errorLabel : ''}`}
+          dateFormat={carbonDateformat}>
+          <DatePickerInput
+            id={`inline-${question.id}`}
+            placeholder={placeholder}
+            labelText={` Date of ${question.label}`} // needs translation
```
Comment (arodidev) [inline src/components/inputs/inline-date/inline-date.component.tsx:80]: Is this something you intended to address?

File: src/components/inputs/inline-date/inline-date.component.tsx [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    return { placeholder: placeholder, carbonDateformat: carbonDateformat };
+  }, []);
+
+  return (
+    <div className={styles.datetime}>
+      <div>
+        <DatePicker
+          datePickerType="single"
+          onChange={onDateChange}
+          className={`${styles.boldedLabel} ${isFieldRequiredError ? styles.errorLabel : ''}`}
```
Comment (arodidev) [inline src/components/inputs/inline-date/inline-date.component.tsx:75]: Maybe we could adopt the use of [classnames](https://github.com/openmrs/openmrs-form-engine-lib/pull/184) for conditionally applying CSS classes.

File: src/components/inputs/text/text.component.tsx [hunk 7]
```diff
@@ -36,16 +40,10 @@ const TextField: React.FC<FormFieldProps> = ({ question, onChange, handler, prev
     }
     if (previousValue !== field.value) {
       onChange(question.id, field.value, setErrors, setWarnings);
-      handler?.handleFieldSubmission(question, field.value, encounterContext);
+      getQuestionValue({ obsDate, question, value: field?.value, handler, encounterContext });
     }
   };
 
-  const setPrevValue = (value: any) => {
```
Comment (arodidev) [inline src/components/inputs/text/text.component.tsx:43]: Thanks for this cleanup.

**Review by arodidev — COMMENTED** [review]
Thanks @hadijahkyampeire left a few minor comments.

File: src/components/inputs/inline-date/inline-date.component.tsx (hunk 5 above)
Comment (hadijahkyampeire) [reply src/components/inputs/inline-date/inline-date.component.tsx:80]: Yes, thank you for catching that, I forgot it.

**Comment by hadijahkyampeire** [comment]
@samuelmale @arodidev I have addressed all issues and added some tests for the inline-date-picker, I used the dropdown component

**Comment by samuelmale** [comment]
Did you see the build failure?

**Comment by hadijahkyampeire** [comment]
Yes @samuelmale the date seems to be changing, I am gonna find a solution.

File: src/components/inputs/content-switcher/content-switcher.component.tsx [hunk 8]
```diff
@@ -9,28 +9,33 @@ import { isTrue } from '../../../utils/boolean-utils';
...
 import { useFieldValidationResults } from '../../../hooks/useFieldValidationResults';
 
+import styles from './content-switcher.scss';
+
 const ContentSwitcher: React.FC<FormFieldProps> = ({ question, onChange, handler, previousValue }) => {
   const { t } = useTranslation();
   const [field] = useField(question.id);
   const { setFieldValue, encounterContext, layoutType, workspaceLayout } = React.useContext(FormContext);
   const { errors, setErrors } = useFieldValidationResults(question);
+  const [obsDate, setObsDate] = useState<Date>();
```
Comment (samuelmale) [inline src/components/inputs/content-switcher/content-switcher.component.tsx:23]: I'm only worried about making these components aware of the "obs" part of things. It introduces some kind of coupling. @hadijahkyampeire Is it possible to have the inlined date component to subscribe to `question.meta.submission` events in a hook and update the value?

Comment (hadijahkyampeire) [reply src/components/inputs/content-switcher/content-switcher.component.tsx:23]: @samuelmale I am not sure I understand what you mean, maybe we shall sync on Monday or you can share more insights about the suggestion.

**Comment by hadijahkyampeire** [comment]
I am gonna close this because I have opened a new PR to address this work https://github.com/openmrs/openmrs-form-engine-lib/pull/285

**Pull request closed without merging**